engine/player/unique_gear_shadowlands.cpp

```diff
@@ -1681,45 +1691,50 @@ void shadowgrasp_totem( special_effect_t& effect )
     {
       buff_t::reset();
 
-      retarget_event = nullptr;
+      last_retarget = 0_s;
     }
 
     void trigger_target_death( const player_t* actor )
     {
-      if ( !check() || !actor->is_enemy() || action->parent->target != actor )
+      auto pet = spawner.pet();
+      if ( !pet || !actor->is_enemy() )
       {
         return;
       }
 
-      item_cd->adjust( cd_adjust );
 
-      if ( !retarget_event && sim->shadowlands_opts.retarget_shadowgrasp_totem > 0_s )
+      if ( pet->target != actor )
       {
-        retarget_event = make_event( sim, sim->shadowlands_opts.retarget_shadowgrasp_totem, [ this ]() {
-          retarget_event = nullptr;
-
-          // Retarget parent action to emulate player "retargeting" during Shadowgrasp
-          // Totem duration to grab more 15 second cooldown reductions
-          {
-            auto new_target = action->parent->select_target_if_target();
-            if ( new_target )
-            {
-              sim->print_debug( "{} action {} retargeting to a new target: {}",
-                                source->name(), action->parent->name(), new_target->name() );
-              action->parent->set_target( new_target );
-            }
-          }
-        } );
+        auto retarget_time = sim->shadowlands_opts.retarget_shadowgrasp_totem;
+        if ( retarget_time == 0_s || sim->current_time() - last_retarget < retarget_time )
+          return;
+        //  Emulate player "retargeting" during Shadowgrasp Totem duration to grab more
+        //  15 second cooldown reductions.
+        //  Don't actually change the taret of the ability, as it will always hit whatever the
+        //  player is hitting.
+        last_retarget = sim->current_time();
       }
+
+      item_cd->adjust( cd_adjust );
     }
   };
 
-  auto buff = buff_t::find( effect.player, "shadowgrasp_totem" );
-  if ( !buff )
+  struct shadowgrasp_totem_t : proc_spell_t
   {
-    buff = make_buff<shadowgrasp_totem_buff_t>( effect );
-    effect.custom_buff = buff;
-  }
+    shadowgrasp_totem_buff_t* buff;
+    shadowgrasp_totem_t( const special_effect_t& e ) :
+      proc_spell_t( "shadowgrasp_totem", e.player, e.player->find_spell( 331523 ) ),
+      buff( new shadowgrasp_totem_buff_t( e ) )
+    {}
+
+    void execute() override
+    {
+      buff->spawner.spawn();
+      buff->trigger();;
+    }
+  };
+
+  effect.execute_action = create_proc_action<shadowgrasp_totem_t>( "shadowgrasp_totem", effect );
```

Comment:
You don't need to create an action just to trigger the buff. You can retain the old code with the sanity check for buff already existing and instead override `shadowgrasp_totem_buff_t::trigger()` to spawn the pet then continue with `buff_t::trigger()`

Reply:
Yes, sorry. Initially I wanted to make sure the pet was spawning before the buff ticks (in case they ever make the buff tick on spawn, now it doesn't), but doing it like you said works as well, I'm just not familiar with all the simc hierarchy.